### PR TITLE
Fix many cases of arithmetic on void pointers

### PR DESCRIPTION
**Pull request by cole-miller**
Currently we use `void *` as a cursor type when parsing and serializing, and this results in many cases of performing arithmetic on void pointers -- technically UB. Change these types to `char *` everywhere.

Signed-off-by: Cole Miller <cole.miller@canonical.com>

### CODE REVIEW

File: src/client/protocol.c [hunk 1]
```diff
@@ -103,7 +103,8 @@ static int peekUint64(struct cursor cursor, uint64_t *val)
 	if (cursor.cap < 8) {
 		return DQLITE_CLIENT_PROTO_ERROR;
 	}
-	*val = ByteFlipLe64(*(uint64_t *)cursor.p);
```
Comment (MathieuBordere) [inline src/client/protocol.c:106]: Where is the arithmetic on a void pointer happening in this case (and similar cases) ?

Comment (cole-miller) [reply src/client/protocol.c:106]: There is no void pointer arithmetic going on, the problem is just that once `cursor.p` is no longer a void pointer the compiler starts to emit a warning about alignment on the cast here. Using memcpy instead gets rid of that.

Comment (cole-miller) [reply src/client/protocol.c:106]: `char *` has no alignment requirements (and can alias with anything) so changing the type can't create any alignment problems.

Comment (cole-miller) [reply src/client/protocol.c:106]: But those situations would already have UB with `void *` right?

Comment (MathieuBordere) [reply src/client/protocol.c:106]: yeah, didn't know void* and char* had same alignment requirements.

**Pull request merged**